src/providers/proxmoxve/cloudconfig.rs

```diff
@@ -0,0 +1,279 @@
+    pub meta_data: ProxmoxVECloudMetaData,
+    pub user_data: Option<ProxmoxVECloudUserData>,
+    pub vendor_data: ProxmoxVECloudVendorData,
+    pub network_config: ProxmoxVECloudNetworkConfig,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudMetaData {
+    #[serde(rename = "instance-id")]
+    pub instance_id: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudUserData {
+    pub hostname: String,
+    pub manage_etc_hosts: bool,
+    pub fqdn: String,
+    pub chpasswd: ProxmoxVECloudChpasswdConfig,
+    pub users: Vec<String>,
+    pub package_upgrade: bool,
+    #[serde(default)]
+    pub ssh_authorized_keys: Vec<String>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudChpasswdConfig {
+    pub expire: bool,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudVendorData {}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudNetworkConfig {
+    pub version: u32,
+    pub config: Vec<ProxmoxVECloudNetworkConfigEntry>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudNetworkConfigEntry {
+    #[serde(rename = "type")]
+    pub network_type: String,
+    pub name: Option<String>,
+    pub mac_address: Option<String>,
+    #[serde(default)]
+    pub address: Vec<String>,
+    #[serde(default)]
+    pub search: Vec<String>,
+    #[serde(default)]
+    pub subnets: Vec<ProxmoxVECloudNetworkConfigSubnet>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudNetworkConfigSubnet {
+    #[serde(rename = "type")]
+    pub subnet_type: String,
+    pub address: Option<String>,
+    pub netmask: Option<String>,
+    pub gateway: Option<String>,
+}
+
+impl ProxmoxVECloudConfig {
+    pub fn try_new(path: &Path) -> Result<Self> {
+        let user_data_file = BufReader::new(File::open(path.join("user-data"))?);
+        let mut user_data = None;
+
+        if let Some(first_line) = user_data_file.lines().next() {
+            if let Ok(first_line) = first_line {
+                if first_line.starts_with("#cloud-config") {
+                    user_data = serde_yaml::from_reader(File::open(path.join("user-data"))?)?;
+                }
+            }
+        }
+
+        if user_data.is_none() {
+            warn!(
+                "user-data does not have the expected header `#cloud-config`, ignoring this file"
+            );
+        }
+
+        Ok(Self {
+            user_data,
+            meta_data: serde_yaml::from_reader(File::open(path.join("meta-data"))?)?,
+            vendor_data: serde_yaml::from_reader(File::open(path.join("vendor-data"))?)?,
+            network_config: serde_yaml::from_reader(File::open(path.join("network-config"))?)?,
+        })
+    }
+}
+
+impl MetadataProvider for ProxmoxVECloudConfig {
+    fn attributes(&self) -> Result<HashMap<String, String>> {
+        let mut out = HashMap::new();
+
+        out.insert(
+            "PROXMOXVE_HOSTNAME".to_owned(),
+            self.hostname()?.unwrap_or_default(),
+        );
```

Comment:
If no hostname is set, I would expect no `PROXMOXVE_HOSTNAME`, not an empty one.